hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java

```diff
@@ -423,6 +425,47 @@ public void deregisterAllListeners(@Nonnull String serviceName, @Nonnull String
         }
     }
 
+    public void cacheListener(Registration registration) {
+        listenerCache.put(registration.getId(), registration.getListener());
+    }
+
+    /**
+     * Adds a {@link Registration} to the corresponding {@link EventServiceSegment} by
+     * avoiding duplicate local registrations.
+     * @implNote There are three code paths for this method: <ol>
+     * <li> {@link #registerListener0} → invokeOnAllMembers({@link RegistrationOperation}).
+     *      In this case, the local registration is already done in {@code registerListener0}
+     *      and thus need to be ignored.
+     * <li> {@code JobProxy.addStatusListener} → {@code AddJobStatusListenerOperation} →
+     *      {@code MasterContext} → {@link #handleAllRegistrations} →
+     *      invokeOnAllMembers({@link RegistrationOperation}). In this case, the local
+     *      registration is new and thus need to be performed.
+     * <li> {@link OnJoinRegistrationOperation}. In this case, no local registration is
+     *      expected. </ol>
+     * The distinction between cases 1 and 2 is made by checking the {@link #listenerCache}
```

Comment:
Can we put `listenerCache` to jet a specific service? Now it seems jet specific logic is leaking here.